### ISSUE TITLE
Badger failure on Windows Subsystem for Linux (fixed on Badger 2.0)

### ISSUE THREAD

**Issue by CMCDragonkai** [smallstep/certificates#156]
### Subject of the issue

I got this bug while trying out step-ca on WSL: https://github.com/dgraph-io/badger/issues/722

```
Error opening database of Type badger with source /tmp/step/db: error opening Badger database: Unable to mmap RDWR log file: exec format error
```

According to the comment at the end, it was fixed when upgrading to badger 2. step-ca still currently uses 1.5.3.

### Your environment

* OS - Windows 10 with Ubuntu WSL

### Steps to reproduce

Use Windows 10 WSL version 1909 OS build 18363.535, get ubuntu wsl. Get step-ca. And try running it after `step ca init --ssh`.

**Comment by strarsis** [comment]
@dopey: Although I don't really need revocations because this should be used for localhost development, it would be still nice to have this fully working on WSL.

**Comment by dopey** [comment]
> @dopey: Although I don't really need revocations because this should be used for localhost development, it would be still nice to have this fully working on WSL. Couldn't badger be used in a way that it autodetects the database file format and offer an optional migration to the user?

We discussed this in the morning, and I'll do my best to take a look this week. Current thinking is try to open the db with v2, if that fails then fallback to v1 and log a message that directs the user to the badger migration docs.



> Edit: The other supported database, bbolt also got issues with latest WSL (go `panic`).
> How has the `ca.json` to be configured to not use database at all?
> I can't just leave the `db` option empty: `Error opening database of Type with source : database not supported`.

Try omitting the db attribute entirely. I've just tried that and it works locally.

**Comment by strarsis** [comment]
*And/or additionally offering a DB type `"type": "badger-v2"`.

**Comment by dopey** [comment]
Yeah, that was my initial idea too. If the method I described above isn't working out then I'll fallback to this.

**Comment by dopey** [comment]
Update on this: I tried to load badger and badger/v2 from the same project and the build fails with the following error:

```
panic: proto: duplicate enum registered: pb.ManifestChange_Operation

goroutine 1 [running]:
github.com/golang/protobuf/proto.RegisterEnum(...)
	/Users/max/pkg/mod/github.com/golang/protobuf@v1.3.1/proto/properties.go:459
github.com/dgraph-io/badger/v2/pb.init.0()
	/Users/max/pkg/mod/github.com/dgraph-io/badger/v2@v2.0.1/pb/pb.pb.go:638 +0x459
```

I believe this is because there is common code being run from the `init.go` for both packages. So, it doesn't look to me like there's an easy way to make both badger and badger/v2 available from the same package. I think we need to choose one or the other.

Short term I can build a branch of `step-ca` that only uses badger/v2. We'll need to come up with a longer term plan for migrating users to v2. Maybe just doing the migration for people ...

Would a branch be useful to people or do you prefer to have specific builds in the github releases?

**Comment by strarsis** [comment]
Could the v1 badger package be dynamically loaded using a [`go` plugin](https://golang.org/pkg/plugin/)?

**Comment by dopey** [comment]
Maybe. We don't use plugins anywhere else in our codebase and I've never used them before. I’m weary of introducing something I don't yet understand.

However, if you'd like to submit a PR that does this, I'd be happy work with you to get it merged.

**Comment by mmalone** [comment]
Could we open an issue with badger to describe the situation and see if they have any guidance?

**Comment by dopey** [comment]
https://github.com/dgraph-io/badger/issues/1208

**Comment by dopey** [comment]
Update on the [issue we opened on Badger](https://github.com/dgraph-io/badger/issues/1208#issuecomment-582868967) is that they are looking into this to see if there's anything they can do to ease the transition :/

**Comment by nicewook** [comment]
Let me check "only badger 2.0 working" interim release is useful

**Comment by dopey** [comment]
I'll build an interim release that supports only badger2 and link it tomorrow.

**Comment by dopey** [comment]
Here's a link to a release built using badger2 -> https://github.com/smallstep/certificates/releases/tag/v0.14.0-rc.4.badger2

**Comment by strarsis** [comment]
@dopey: Just pulled the latest image from DockerHub (auto-build, nice!).
I get an error message from traefik, apparently after a restart of traefik and/or the step-ca server: `urn:ietf:params:acme:error:serverInternal :: error creating order: error storing order IDs for account mv4GuMH6U2shGwN7r26AM97qc2UaxcQf; order IDs changed since last read, url: " providerName=localca.acme routerName=traefik-sec@docker`

**Comment by nicewook** [comment]
> Here's a link to a release built using badger2 -> https://github.com/smallstep/certificates/releases/tag/v0.14.0-rc.4.badger2

Thank you @dopey I just tested my Ubuntu in Windows 10. 
and I go through the link https://smallstep.com/docs/getting-started/

**Comment by gclawes** [comment]
I'm seeing this on the latest release `Smallstep CA/0.14.2 (linux/arm)`:
```
root@ca:~# step-ca --version
Smallstep CA/0.14.2 (linux/arm)
Release Date: 2020-04-10 03:50 UTC
root@ca:~# step-ca $(step path)/config/ca.json
Error opening database of Type badger with source /root/.step/db: error opening Badger database: Unable to mmap RDWR log file: invalid argument
```

**Comment by dopey** [comment]
@gclawes was a previous version of `step` working for you on linux/arm? Want to make sure this isn't a regression. We just introduced those builds to the releases with this version and I haven't actually verified that they work on those systems. 

I just tested on linux/ubuntu and darwin and don't have any trouble opening the DB. 

If this is, indeed, an issue with linux/arm not supporting badger v1, then we should be able to fix this in the 0.15 release, which should be a quick follow. Badger recently pushed a fix allowing the use of both badger v1 and v2 simultaneously, which unblocks us being able to support badger v2. 

I'm happy to create a release that uses badgerV2 if you'd be willing to try that out to see if it fixes your issue.

**Comment by gclawes** [comment]
This is the first version I am testing on ARM.

I see the same with `0.14.0-dev` and `0.14.0-rc.16` (the first ARM release):
```
root@ca:~# step-ca --version
Smallstep CA/0.14.0-dev (linux/arm)
Release Date: 2020-04-09 16:45 UTC
root@ca:~# step-ca $(step path)/config/ca.json
Error opening database of Type badger with source /root/.step/db: error opening Badger database: Unable to mmap RDWR log file: invalid argument
```
```
root@ca:~# step-ca --version
Smallstep CA/0.14.0-rc.16 (linux/arm)
Release Date: 2020-04-08 22:29 UTC
root@ca:~# step-ca $(step path)/config/ca.json
Error opening database of Type badger with source /root/.step/db: error opening Badger database: Unable to mmap RDWR log file: invalid argument
```

**Comment by gclawes** [comment]
I'd be happy to test anything you want.My testin

**Comment by dopey** [comment]
Made a badger2 release - https://github.com/smallstep/certificates/releases v014.3-rc.1.badger2. Let me know if that works (or doesn't).

**Comment by gclawes** [comment]
I'm getting this on my armv7 box:
```
root@ca:~# step version
Smallstep CLI/0.14.2 (linux/arm)
Release Date: 2020-04-10 03:39 UTC
root@ca:~# step-ca $(step path)/config/ca.json
Error opening database of Type badger with source /root/.step/db: error opening database: error opening Badger database: manifest has unsupported version: 4 (we support 7).
Please see https://github.com/dgraph-io/badger/blob/master/README.md#i-see-manifest-has-unsupported-version-x-we-support-y-error on how to fix this.
```

**Comment by gclawes** [comment]
Removing /root/.step and re-running `step ca init` fixed this:
```
root@ca:~# step ca init
✔ What would you like to name your new PKI? (e.g. Smallstep): ca.lan
✔ What DNS names or IP addresses would you like to add to your new CA? (e.g. ca.smallstep.com[,1.1.1.1,etc.]): ca.lan,10.10.0.65
✔ What address will your new CA listen at? (e.g. :443): :443
✔ What would you like to name the first provisioner for your new CA? (e.g. you@smallstep.com): root@ca.lan
✔ What do you want your password to be? [leave empty and we'll generate one]:
✔ Password: w1L>jIlkdui12{Mg6g)6pK5NQE(bcD<&

Generating root certificate...
all done!

Generating intermediate certificate...
all done!

✔ Root certificate: /root/.step/certs/root_ca.crt
✔ Root private key: /root/.step/secrets/root_ca_key
✔ Root fingerprint: ebc95ea862c46bf6dd37d932827c8b71cb5aadc812374cc3e9af7aa7d35388f2
✔ Intermediate certificate: /root/.step/certs/intermediate_ca.crt
✔ Intermediate private key: /root/.step/secrets/intermediate_ca_key
✔ Database folder: /root/.step/db
✔ Default configuration: /root/.step/config/defaults.json
✔ Certificate Authority configuration: /root/.step/config/ca.json

Your PKI is ready to go. To generate certificates for individual services see 'step help ca'.

FEEDBACK 😍 🍻
      The step utility is not instrumented for usage statistics. It does not
      phone home. But your feedback is extremely valuable. Any information you
      can provide regarding how you’re using `step` helps. Please send us a
      sentence or two, good or bad: feedback@smallstep.com or join
      https://gitter.im/smallstep/community.
root@ca:~# step-ca $(step path)/config/ca.json
badger 2020/04/13 23:29:44 INFO: All 0 tables opened in 0s
Please enter the password to decrypt /root/.step/secrets/intermediate_ca_key:
2020/04/13 23:30:13 Serving HTTPS on :443 ...
```

I originally had over-written an existing `/root/.step` directory, which resulted in the error above:
```
root@ca:~# step ca init
✔ What would you like to name your new PKI? (e.g. Smallstep): ca.lan
✔ What DNS names or IP addresses would you like to add to your new CA? (e.g. ca.smallstep.com[,1.1.1.1,etc.]): ca.lan,10.10.0.65
✔ What address will your new CA listen at? (e.g. :443): :443
✔ What would you like to name the first provisioner for your new CA? (e.g. you@smallstep.com): root@ca.lan
✔ What do you want your password to be? [leave empty and we'll generate one]:
✔ Password: $\n9s!>6/:Zw8HF]N6EY>>'mjf4]LpdA

Generating root certificate...
✔ Would you like to overwrite /root/.step/certs/root_ca.crt [y/n]: y
✔ Would you like to overwrite /root/.step/secrets/root_ca_key [y/n]: y
all done!

Generating intermediate certificate...
✔ Would you like to overwrite /root/.step/certs/intermediate_ca.crt [y/n]: y
✔ Would you like to overwrite /root/.step/secrets/intermediate_ca_key [y/n]: y
all done!

✔ Root certificate: /root/.step/certs/root_ca.crt
✔ Root private key: /root/.step/secrets/root_ca_key
✔ Root fingerprint: 8883cf83d996b6fd264c054fb04c3c3ec7b0d9263d9c71de324d24c1810c92fd
✔ Intermediate certificate: /root/.step/certs/intermediate_ca.crt
✔ Intermediate private key: /root/.step/secrets/intermediate_ca_key
✔ Would you like to overwrite /root/.step/config/ca.json [y/n]: y
✔ Would you like to overwrite /root/.step/config/defaults.json [y/n]: y
✔ Database folder: /root/.step/db
✔ Default configuration: /root/.step/config/defaults.json
✔ Certificate Authority configuration: /root/.step/config/ca.json

Your PKI is ready to go. To generate certificates for individual services see 'step help ca'.

FEEDBACK 😍 🍻
      The step utility is not instrumented for usage statistics. It does not
      phone home. But your feedback is extremely valuable. Any information you
      can provide regarding how you’re using `step` helps. Please send us a
      sentence or two, good or bad: feedback@smallstep.com or join
      https://gitter.im/smallstep/community.
root@ca:~# step-ca $(step path)/config/ca.json
Error opening database of Type badger with source /root/.step/db: error opening database: error opening Badger database: manifest has unsupported version: 4 (we support 7).
Please see https://github.com/dgraph-io/badger/blob/master/README.md#i-see-manifest-has-unsupported-version-x-we-support-y-error on how to fix this.
```
